### PR TITLE
CSVS-57 CircleQueue was added

### PR DESCRIPTION
**Pull request by MedvedevBW**

### CODE REVIEW

File: src/main/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreator.scala [hunk 1]
```diff
@@ -151,6 +142,10 @@ class CloudStackTaskCreator(settings: CloudStackTaskCreator.Settings) {
         throw new CloudStackFatalException(e.toString)
     }
   }
+
+  protected def getClient(endpoint: String): ApacheCloudStackClient = {
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreator.scala]: createClient()

File: src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+package com.bwsw.cloudstack.vault.server.common
+
+import org.scalatest.{FlatSpec, Matchers}
+
+class CircleQueueTestSuite extends FlatSpec with Matchers {
+  val firstElement = "1"
+  val secondElement = "2"
+  val thirdElement = "3"
+
+  "moveElementToEnd" should "move the first element to end of the queue" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala]: "moveElementToEnd" should "move the first element to the end of the queue"

File: src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    val expectedResult = List(secondElement, thirdElement, firstElement)
+    val queue = new CircleQueue[String](List(firstElement, secondElement, thirdElement))
+
+    queue.moveElementToEnd(firstElement)
+
+    assert(expectedResult == queue.getElements)
+  }
+
+  "moveElementToEnd" should "not change queue if element for move is not first" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala]: "moveElementToEnd" should "not change a queue if an element is not first"

**Review by placccebo — CHANGES_REQUESTED** [review]

File: src/test/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreatorTestSuite.scala [hunk 4]
```diff
@@ -128,16 +127,15 @@ class CloudStackTaskCreatorTestSuite extends FlatSpec with TestData with BaseTes
   "createRequest" should "create request" in {
 
     val response = "testResponse"
-    val urlClient = "http://127.0.0.1:8080/client/api/2"
 
     val createRequest = PrivateMethod[String]('createRequest)
 
     val cloudStackTaskCreator = new CloudStackTaskCreator(cloudStackTaskCreatorSettings) {
-      override val endpointQueue = new CircleQueue[String](cloudStackTaskCreatorSettings.endpoints.toList)
+      override val endpointQueue = getEndPointQueue(cloudStackTaskCreatorSettings.endpoints.toList)
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreatorTestSuite.scala]: getEndpointQueue

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  fillRangeMap(0, lengthOfArray, 0)
+
+  def getElement: T = elements(gradeScaleElements.get(r.nextInt(sumOfNumbers)))
+
+  def getElements: List[T] = elements
+
+  def moveElementToEnd(element: T): Unit = synchronized {
+    val currentIndex = elements.indexOf(element)
+    if(currentIndex != -1) {
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: add a space between "if" and condition: if (currentIndex != -1) ....

Comment (placccebo) [reply src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: check the code below

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  * (lengthOfList - indexOfElement)/sumOfAllNumbersOfList
+  * where:
+  * "lengthOfList" - length of initialization list with elements
+  * "indexOfElement" - index of element in the list
+  * "sumOfAllNumbersOfList" - sum of numbers from 1 to "lengthOfList"
+  */
+class WeightedQueue[T](private val elementList: List[T]) {
+  protected val r = new Random
+
+  private val lengthOfArray = elementList.length
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: why does the name contain "array"? is it a list or array?

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  def getElements: List[T] = elements
+
+  def moveElementToEnd(element: T): Unit = synchronized {
+    val currentIndex = elements.indexOf(element)
+    if(currentIndex != -1) {
+      if (currentIndex != lengthOfArray - 1) {
+        elements = elements.take(currentIndex) ++ elements.takeRight(lengthOfArray - currentIndex - 1) :+ element
+      }
+    } else {
+      throw new IllegalArgumentException(s"Element: $element is not included in the initialization list: $elementList")
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: Requested element doesn't exist: $element

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  * "lengthOfList" - length of initialization list with elements
+  * "indexOfElement" - index of element in the list
+  * "sumOfAllNumbersOfList" - sum of numbers from 1 to "lengthOfList"
+  */
+class WeightedQueue[T](private val elementList: List[T]) {
+  protected val r = new Random
+
+  private val lengthOfArray = elementList.length
+  private val sumOfNumbers = getSumOfNumbers(0, lengthOfArray)
+  private val gradeScaleElements: RangeMap[Integer, Int] = TreeRangeMap.create()
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: elementIndexByPossibility

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+  private var elements = elementList
+
+  fillRangeMap(0, lengthOfArray, 0)
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: rename method

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  private def fillRangeMap(startOfRange: Int, range: Int, index: Int): Unit = {
+    if (range != 0) {
+      gradeScaleElements.put(Range.closed(startOfRange, startOfRange + range - 1), index)
+      fillRangeMap(startOfRange + range, range - 1, index + 1)
+    }
+  }
+
+  private def getSumOfNumbers(acc: Int, lenght: Int): Int = {
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: getSum() = N * (first + last) / 2

**Review by placccebo — CHANGES_REQUESTED** [review]

File: src/main/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreator.scala (hunk 1 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreator.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala (hunk 2 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala (hunk 3 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/vault/server/common/CircleQueueTestSuite.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala (hunk 6 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: List

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala (hunk 8 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala (hunk 9 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala (hunk 5 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala (hunk 7 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala (hunk 10 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: fixed

File: src/test/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreatorTestSuite.scala (hunk 4 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/vault/server/cloudstack/util/CloudStackTaskCreatorTestSuite.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala [hunk 11]
```diff
@@ -33,41 +33,33 @@ import scala.util.Random
 class WeightedQueue[T](private val elementList: List[T]) {
   protected val r = new Random
 
-  private val lengthOfArray = elementList.length
-  private val sumOfNumbers = getSumOfNumbers(0, lengthOfArray)
-  private val gradeScaleElements: RangeMap[Integer, Int] = TreeRangeMap.create()
+  private val lengthOfList = elementList.length
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/common/WeightedQueue.scala]: just length or size

**Review by placccebo — CHANGES_REQUESTED** [review]

**Review by placccebo — APPROVED** [review]

**Pull request merged**